### PR TITLE
[#3530] Create tests for FeedbackResponseCommentsLogicTest::isNameVisibleToUser

### PR DESCRIPTION
**Pull request by hhdqirui**
Part of #3530 

**Outline of Solution**
For the ease of testing, a data bundle json file is created specifically for `FeedbackResponseCommentsLogicTest.java`. This data bundle is similar to the typical data bundle except that a student, `student6InCourse1` is added and this student is in the same team as `student5InCourse1`.

### CODE REVIEW

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java [hunk 1]
```diff
@@ -346,6 +355,118 @@ public void testGetFeedbackResponseCommentsForQuestionInSection_withSectionName_
...
+        // test user's participant type is in the list that comment giver's name is shown to
+        comment = dataBundle.feedbackResponseComments.get("comment1FromT1C1ToR1Q2S1C1");
+        relatedResponse = dataBundle.feedbackResponses.get("response1ForQ2S1C1");
+        relatedResponse.giver = "student1InCourse1@gmail.tmt";
+
+        ______TS("success: comment is only visible to instructors");
+        comment.commentGiver = "student6@course1.tmt";
+        comment.commentGiverType = STUDENTS;
+        comment.showGiverNameTo = Arrays.asList(INSTRUCTORS);
+        assertTrue(frcLogic.isNameVisibleToUser(comment, relatedResponse, "instructor2@course1.tmt", roster));
```
Comment (daongochieu2810) [inline src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:388]: e.g "instructor2@course1.tmt" can be a constant as its being reused in a lot of places

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java [hunk 2]
```diff
@@ -30,7 +39,7 @@ protected void prepareTestData() {
 
     @BeforeMethod
     public void refreshTestData() {
-        dataBundle = getTypicalDataBundle();
+        dataBundle = loadDataBundle("/FeedbackResponseCommentsLogicTest.json");
         removeAndRestoreTypicalDataBundle();
```
Comment (daongochieu2810) [inline src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:43]: Is this necessary if we are not using `getTypicalDataBundle();`

File: src/test/resources/data/FeedbackResponseCommentsLogicTest.json [hunk 3]
```diff
@@ -0,0 +1,1709 @@
+{
+  "accounts": {
```
Comment (daongochieu2810) [inline src/test/resources/data/FeedbackResponseCommentsLogicTest.json:2]: Why do we need a separate data file for this test? If you want to add new data for the new test then load the typical data and load the new data, then merge them

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]

File: src/test/resources/data/FeedbackResponseCommentsLogicTest.json (hunk 3 above)
Comment (hhdqirui) [reply src/test/resources/data/FeedbackResponseCommentsLogicTest.json:2]: The reason for not using `TypicalDataBundle` is that adding new data to `TypicalDataBundle` will break a lot of other tests for some reason.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 2 above)
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:43]: `TypicalDataBundle` does not have the desired data for testing and add new data to it will break many other tests.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java [hunk 4]
```diff
@@ -346,6 +355,118 @@ public void testGetFeedbackResponseCommentsForQuestionInSection_withSectionName_
         assertEquals(0, comments.size());
     }
 
+    @Test
+    public void testIsNameVisibleToUser() {
+        FeedbackResponseCommentAttributes comment = dataBundle.feedbackResponseComments.get("comment1FromT1C1ToR1Q1S1C1");
```
[the remark this answers is not in the document]
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java]: Thank you for the advice. It is updated now.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 2 above)
Comment (daongochieu2810) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:58]: I mean to suggest that in the case of not using `TypicalDataBundle`, `removeAndRestoreTypicalDataBundle()` would not be necessary. You can check the method declaration here
Screenshot from 2021-06-11 22-55-21

File: src/test/resources/data/FeedbackResponseCommentsLogicTest.json (hunk 3 above)
Comment (daongochieu2810) [reply src/test/resources/data/FeedbackResponseCommentsLogicTest.json:2]: @hhdqirui I think you can take a look at `teammates/logic/core/FeedbackResponsesLogicTest.java` to check how it is using both `typicalDataBundle` and extra data bundle

Comment (hhdqirui) [reply src/test/resources/data/FeedbackResponseCommentsLogicTest.json:2]: Thank you for the suggestion. I will look into that.

**Comment by hhdqirui** [comment]
Hi @daongochieu2810. I have removed the `FeedbackResponseCommentsLogicTest.json` file and revert it to use `TypicalDataBundle`. A new data bundle file `CourseRosterDataBundle.json` is added for testing of this class. The new data bundle file only contains `instructors` and `students` from the typical data bundle with additional student added. It is used together with `TypicalDataBundle` for testing.

It is ready for review now and would like to get your advice on this. Thank you!

File: src/test/resources/data/CourseRosterDataBundle.json [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+          "canviewsessioninsection": false,
+          "canmodifysession": false,
+          "canmodifystudent": false,
+          "canmodifyinstructor": false
+        },
+        "sectionLevel": {},
+        "sessionLevel": {}
+      }
+    },
+    "instructorNotYetJoinCourse1": {
```
Comment (daongochieu2810) [inline src/test/resources/data/CourseRosterDataBundle.json:75]: I dont think this instructor is needed here

File: src/test/resources/data/CourseRosterDataBundle.json [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+          "canviewsessioninsection": true,
+          "canmodifysession": true,
+          "canmodifystudent": true,
+          "canmodifyinstructor": true
+        },
+        "sectionLevel": {},
+        "sessionLevel": {}
+      }
+    },
+    "instructorWithOnlyOneSampleCourse": {
```
Comment (daongochieu2810) [inline src/test/resources/data/CourseRosterDataBundle.json:315]: this also

File: src/test/resources/data/CourseRosterDataBundle.json [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+          "canviewsessioninsection": true,
+          "canmodifysession": true,
+          "canmodifystudent": true,
+          "canmodifyinstructor": true
+        },
+        "sectionLevel": {},
+        "sessionLevel": {}
+      }
+    },
+    "instructorOfArchivedCourse": {
```
Comment (daongochieu2810) [inline src/test/resources/data/CourseRosterDataBundle.json:339]: same here

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]
The changes look a lot better now. But I think the new test data file still contains some unused data, maybe we can remove that

File: src/test/resources/data/CourseRosterDataBundle.json (hunk 7 above)
Comment (hhdqirui) [reply src/test/resources/data/CourseRosterDataBundle.json:339]: All these instructors are from the typical data bundle. If we just use typical data bundle for testing, these instructors would also be part of the testing data.

**Comment by hhdqirui** [comment]
Hi @daongochieu2810. I have remove those unused data and it is ready for review.

**Review by daongochieu2810 — APPROVED** [review]
lgtm!

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java [hunk 8]
```diff
@@ -23,6 +33,23 @@
     private static final FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();
     private static final FeedbackResponsesLogic frLogic = FeedbackResponsesLogic.inst();
 
+    private static final String COMMENT1_FROM_T1C1_TO_R1Q1S1C1 = "comment1FromT1C1ToR1Q1S1C1";
```
Comment (rrtheonlyone) [inline src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:36]: Why all these static final variables? 

I know @daongochieu2810 mentioned to abstract out as a comment and that is a reasonable suggestion. 

The reason I don't like this is because:
1. Look at the other tests, none of them have this 
2. It is adding unnecessary clutter - there is nothing wrong with repeating the email in a test function.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java [hunk 9]
```diff
@@ -31,6 +58,7 @@ protected void prepareTestData() {
     @BeforeMethod
     public void refreshTestData() {
         dataBundle = getTypicalDataBundle();
+        courseRosterDataBundle = loadDataBundle("/CourseRosterDataBundle.json");
```
Comment (rrtheonlyone) [inline src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:61]: Are you sure that the typical data bundle is insufficient to test this function?

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java [hunk 10]
```diff
@@ -346,6 +374,118 @@ public void testGetFeedbackResponseCommentsForQuestionInSection_withSectionName_
...
+        assertFalse(frcLogic.isNameVisibleToUser(comment, relatedResponse, STUDENT5_IN_COURSE1_EMAIL, roster));
+        assertFalse(frcLogic.isNameVisibleToUser(comment, relatedResponse, STUDENT1_IN_COURSE1_EMAIL, roster));
+        assertFalse(frcLogic.isNameVisibleToUser(comment, relatedResponse, STUDENT3_IN_COURSE1_EMAIL, roster));
+        assertFalse(frcLogic.isNameVisibleToUser(comment, relatedResponse, STUDENT1_IN_COURSE2_EMAIL, roster));
+
+        ______TS("success: comment is only visible to response recipient");
+        comment.commentGiver = STUDENT4_IN_COURSE1_EMAIL;
+        comment.commentGiverType = STUDENTS;
+        comment.showGiverNameTo = Arrays.asList(RECEIVER);
+        assertFalse(frcLogic.isNameVisibleToUser(comment, relatedResponse, INSTRUCTOR2_IN_COURSE1_EMAIL, roster));
```
Comment (rrtheonlyone) [inline src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:417]: There is no reason to test so many false cases. Check if they are really testing different behaviour and if so, leave it in.

Refer to "Equivalence Partitions" sections here: https://nus-cs2103-ay1920s2.github.io/website/se-book-adapted/chapters/testCaseDesign.html

```
By dividing possible inputs into equivalence partitions we can:
- avoid testing too many inputs from one partition. Testing too many inputs from the same partition is unlikely to find new bugs. This increases the efficiency of testing by reducing redundant test cases.
- ensure all partitions are tested. Missing partitions can result in bugs going unnoticed. This increases the effectiveness of testing by increasing the chance of finding bugs.
```

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Good Work, some minor comments below:

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 8 above)
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:36]: So changing them to local variable of the test function instead of having them as static variable of the class would solve the problem here?

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 9 above)
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:61]: I would like to have a feedback response of student from one team to student of another team. typical data bundle have one such response which is between student 2 in course 1 and student 5 in course 1. However, student 2 in course 1 is the same student as student 2 in course 2. When the data bundle is loaded, this student will be treated as student 2 in course 2 instead of student 2 in course 1 and the team that this student is in is the team in course 2. It creates some ambiguityies here.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 10 above)
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:417]: From line 403 and below, each of the test segmentation is testing for each case in `FeedbackResponseCommentsLogic::isFeedbackParticipantNameVisibleToUser`. And each of the assertion under each test segmentation is testing for one of the following groups: instructor, response receiver, response receiver's teammate, response giver, response giver's teammate and other students. I believe here each partition is referring to each case? And in each partition, testing for all the groups of people seems to be necessary and for each group of people I only use one input to avoid inefficiency.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 8 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:36]: This is arguable but I would prefer that yes. What are your thoughts?

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 9 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:61]: I am not sure what you mean by the above. One student can be in multiple courses. How about you paste to me which cases the above affects and I try to rewrite it purely using the typical bundle?

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 10 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:417]: That is an interesting way to look at. If you notice these lines:
```
        for (FeedbackParticipantType type : showNameTo) {
            switch (type) {
            case INSTRUCTORS:
                if (roster.getInstructorForEmail(userEmail) != null) {
                    return true;
                }
                break;
            case OWN_TEAM_MEMBERS:
                if (responseGiverTeam.equals(currentUserTeam)) {
                    return true;
                }
                break;
                ....
 ```
 
They only return True. So by testing all the true branches, I would have tested these cases. I would then need exactly one test where it matches none of the branches to return False.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 8 above)
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:36]: Now I also think that it would be better to change it to local variables of the test function since no other test functions use them.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 10 above)
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:417]: Ok I see. I will make changes for that.

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 9 above)
Comment (hhdqirui) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:46]: I have created a new test case to illustrate my point. It is from line 471 to 478.

**Comment by hhdqirui** [comment]
After updating with the base branch, the test case I mentioned in https://github.com/TEAMMATES/teammates/pull/11171#discussion_r669403294 is now at line 475 to 483.

**Comment by rrtheonlyone** [comment]
@hhdqirui regardless of the typical bundle vs new bundle issue, your code should be passing CI. 

Leave the bundle issue to me. Please get the PR ready for review without that when you get the time!

**Comment by hhdqirui** [comment]
Hi @rrtheonlyone. The test case that I mentioned in https://github.com/TEAMMATES/teammates/pull/11171#discussion_r669403294 is pasted below. The test case is supposed to be true but it is not false.

```java
______TS("Test comment is visible to response giver's team member. Response giver is student 2 in course 1."
                + "Test whether comment is visible to student 1 in course 1 who are in the same team as student 2 in "
                + "course 1. The team that these two students are in is Team 1.1</td></div>'\"");
roster = new CourseRoster(new ArrayList<>(dataBundle.students.values()), new ArrayList<>(dataBundle.instructors.values()));
FeedbackResponseAttributes response = dataBundle.feedbackResponses.get("response1ForQ2S1C1");
response.setGiver("student2InCourse1@gmail.tmt");
comment.setShowGiverNameTo(Arrays.asList(OWN_TEAM_MEMBERS));
assertTrue(frcLogic.isNameVisibleToUser(comment, response, "student1InCourse1@gmail.tmt", roster));
```

**Comment by ypinhsuan** [comment]
@rrtheonlyone for your kind review if you have time

File: src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java (hunk 9 above)
Comment (rrtheonlyone) [reply src/test/java/teammates/logic/core/FeedbackResponseCommentsLogicTest.java:50]: Upon further thought, I guess it is not a huge harm creating this bundle here.

Even if it seems possible to refactor this to use purely the typical data bundle, it would require some re-engineering of the test. It may not be worth the effort required.

**Review by rrtheonlyone — APPROVED** [review]
LGTM 💯 I think we are good here, thanks @hhdqirui
